src/View/mainpage/RenameROIWindow.py

```diff
@@ -101,18 +105,58 @@ def on_text_edited(self, text):
                     self.feedback_text.setText("Entered text exists but should be in capitals")
 
     def on_rename_clicked(self):
-        new_name = self.input_field.text()
-        new_dataset = ROI.rename_roi(self.rtss, self.roi_id, new_name)
-        # TODO rather than save this so a file straight away, the dataset is sent as part of the signal further
-        # up towards the main page and then 'replaces' the current rtss. at some point the user is given the
-        # option of saving this new rtss file. when the rtss file is replaced it also recalculates the rois and
-        # reload the structure widget's structures.
-        print()
-        self.rename_signal.emit((new_dataset, {"rename": [self.roi_name, new_name]}))
+        self.new_name = self.input_field.text()
+        progress_window = RenameROIProgressWindow(self, QtCore.Qt.WindowTitleHint)
+        progress_window.signal_roi_renamed.connect(self.on_roi_renamed)
+        progress_window.start_renaming(self.rtss, self.roi_id, self.new_name)
+        progress_window.show()
+
+    def on_roi_renamed(self, new_rtss):
+        self.rename_signal.emit((new_rtss, {"rename": [self.roi_name, self.new_name]}))
+        QtWidgets.QMessageBox.about(self, "Saved", "Region of interest successfully renamed!")
         self.close()
 
     def on_ROI_clicked(self):
         clicked_ROI = self.list_of_ROIs.currentItem()
         # Excluding headers from being clicked.
         if not str(clicked_ROI.text()).startswith("------------Standard"):
             self.input_field.setText(str(clicked_ROI.text()))
+
+
+class RenameROIProgressWindow(QtWidgets.QDialog):
```

Comment:
OK.  Now that there are three classes that share a lot, it's likely that creating a base class and deriving would be worthwhile.

Reply:
I'll spend a bit more time on this, because I think you're probably right.